libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -0,0 +1,719 @@
+{
+    MQTTPublishState_t calculatedState = MQTTInvalidState;
+    switch( qos )
+    {
+        case MQTTQoS0:
+            if( ( idStatus == ID_UNKNOWN ) && ( packetType == MQTT_PUBLISH ) )
+            {
+                if( opType == MQTT_RESERVE )
+                {
+                    calculatedState = MQTTPublishSend;
+                }
+                else /* Sent or received. */
+                {
+                    calculatedState = MQTTPublishDone;
+                }
+            }
+            /* Anything else is an invalid state. */
+            break;
+        case MQTTQoS1:
+            calculatedState = _calculateQoS1State( packetType, opType, idStatus );
+            break;
+        case MQTTQoS2:
+            calculatedState = _calculateQoS2State( packetType, opType, idStatus );
+            break;
+        default:
+            break;
+    }
+    return calculatedState;
+}
+
+MQTTStatus_t _MQTT_UpdateState( MQTTContext_t * pMqttContext,
+                                uint16_t packetId,
+                                MQTTPublishType_t packetType,
+                                MQTTStateOperation_t opType,
+                                MQTTQoS_t qos )
+{
+    MQTTStatus_t status = MQTTIllegalState;
+    bool isPublishOutgoing = _isPublishOutgoing( packetType, opType );
+    bool shouldStoreRecord = false;
+    bool shouldDeleteRecord = false;
+    bool isTransitionValid = false;
+    MQTTPacketIDState_t idStatus = ID_INVALID;
+    MQTTPublishState_t currentState = MQTTInvalidState;
+    MQTTPublishState_t newState = MQTTInvalidState;
+    MQTTQoS_t foundQoS = MQTTInvalidQoS;
+    uint16_t recordIndex = MQTT_MAX_QUEUED_PUBLISH_MESSAGES;
+    MQTTPubAckInfo_t * records = NULL;
+
+    if( isPublishOutgoing )
+    {
+        records = pMqttContext->outgoingPublishRecords;
+    }
+    else
+    {
+        records = pMqttContext->incomingPublishRecords;
+    }
+
+    recordIndex = _findInRecord( records,
+                                 MQTT_MAX_QUEUED_PUBLISH_MESSAGES,
+                                 packetId,
+                                 &idStatus,
+                                 &foundQoS,
+                                 &currentState );
+
+    /* Reconcile the supplied QoS with the one in the record. */
+    status = _resolveQoS( qos, &foundQoS );
+
+    if( status == MQTTSuccess )
+    {
+        newState = _MQTT_CalculateState( packetType, opType, foundQoS, idStatus );
+
+        if( newState == MQTTInvalidState )
+        {
+            status = MQTTIllegalState;
+            /* Test if collision. */
+            if( opType == MQTT_RESERVE && idStatus == ID_KNOWN )
+            {
+                status = MQTTStateCollision;
+            }
+        }
+    }
+    if( status == MQTTSuccess )
+    {
+        if( foundQoS != MQTTQoS0 )
+        {
+            shouldStoreRecord = ( newState == MQTTPublishSend ) ||
+                                ( newState == MQTTPubAckSend ) ||
```

Comment:
Is storing a record required for incoming QoS 1 PUBLISH?

Reply:
Similar comment for the outgoing QoS 1 PUBLISH. If the stored record update is not notified to the user on state transition, then the behavior is the same as just notifying a PUBACK (without maintaining any record).

Reply:
After discussion offline, yes, it's desired to store the record for QoS 1 so that a session can be resumed.